app/addons.go

```diff
@@ -102,34 +116,124 @@ func (a *App) AddAddonHandler(subject, reply string, request *qms.AddAddonReques
 func (a *App) ListAddonsHandler(subject, reply string, request *qms.NoParamsRequest) {
 	var err error
 
-	log := log.WithField("context", "list addons")
+	ctx, span := qmsinit.InitNoParamsRequest(request, subject)
+	defer span.End()
 
+	log := log.WithField("context", "list addons")
+	sendError := a.sendAddonListResponseError(reply, log)
 	response := qmsinit.NewAddonListResponse()
+	d := db.New(a.db)
 
-	sendError := func(ctx context.Context, response *qms.AddonListResponse, err error) {
+	results, err := d.ListAddons(ctx)
+	if err != nil {
+		sendError(ctx, response, err)
+		return
+	}
+
+	for _, addon := range results {
+		response.Addons = append(response.Addons, addon.ToQMSType())
+	}
+
+	if err = a.client.Respond(ctx, reply, response); err != nil {
+		log.Error(err)
+	}
+}
+
+func (a *App) UpdateAddonHandler(subject, reply string, request *qms.UpdateAddonRequest) {
+	var err error
+
+	log := log.WithField("context", "update addon")
+
+	response := qmsinit.NewAddonResponse()
+
+	sendError := func(ctx context.Context, response *qms.AddonResponse, err error) {
 		log.Error(err)
 		response.Error = serrors.NatsError(ctx, err)
 		if err = a.client.Respond(ctx, reply, response); err != nil {
 			log.Error(err)
 		}
 	}
 
-	ctx, span := qmsinit.InitNoParamsRequest(request, subject)
+	ctx, span := qmsinit.InitUpdateAddonRequest(request, subject)
 	defer span.End()
 
 	d := db.New(a.db)
 
-	results, err := d.ListAddons(ctx)
+	if request.Addon.Uuid == "" {
+		sendError(ctx, response, errors.New("uuid must be set in the request"))
+		return
+	}
+
+	updateAddon := db.NewUpdateAddonFromQMS(request)
+
+	result, err := d.UpdateAddon(ctx, updateAddon)
 	if err != nil {
 		sendError(ctx, response, err)
 		return
 	}
 
-	for _, addon := range results {
-		response.Addons = append(response.Addons, addon.ToQMSType())
+	response.Addon = result.ToQMSType()
+
+	if err = a.client.Respond(ctx, reply, response); err != nil {
+		log.Error(err)
+	}
+}
+
+func (a *App) DeleteAddonHandler(subject, reply string, request *requests.ByUUID) {
+	var err error
+
+	log := log.WithField("context", "delete addon")
+
+	response := qmsinit.NewAddonResponse()
+
+	sendError := func(ctx context.Context, response *qms.AddonResponse, err error) {
+		log.Error(err)
+		response.Error = serrors.NatsError(ctx, err)
+		if err = a.client.Respond(ctx, reply, response); err != nil {
+			log.Error(err)
+		}
+	}
+
+	ctx, span := reqinit.InitByUUID(request, subject)
+	defer span.End()
+
+	d := db.New(a.db)
+
+	if err = d.DeleteAddon(ctx, request.Uuid); err != nil {
+		sendError(ctx, response, err)
+		return
+	}
+
+	response.Addon = &qms.Addon{
+		Uuid: request.Uuid,
 	}
 
 	if err = a.client.Respond(ctx, reply, response); err != nil {
 		log.Error(err)
 	}
 }
+
+func (a *App) ToggleAddonPaidHandler(subject, reply string, request *requests.ByUUID) {
```

Comment:
I'm not sure if this will cause issues, but I guess this isn't idempotent. Showing my unfamiliarity with NATS/the other upstream and downstream stuff, but would it be better to have this be a more explicit "set paid"/"set unpaid" model instead of a toggle?

Reply:
I went back and forth on this before implementing it. Setting the paid status can be done explicitly through the update functionality, so the idempotent version of this exists. I'm honestly kinda leaning towards removing this stuff/not exposing it through Terrain because it seemed redundant after adding the ability to update an add-on.